### PR TITLE
LoadAware Scheduling

### PR DESCRIPTION
**Pull request by swarnim-vyas**
This PR adds load aware scheduling.
It utilizes 2 data structures:
01) Priority Queue
02) Dictionary

When a worker registers itself with scheduler, we update the priority queue with max capacity of the worker and we update the dictionary with 0 task finished count capacity. 

Please note max capacity of the worker is computed using available threads and generation of hardware.

Once a task comes in, we try to get worker from the priority queue and then if none are available there we try to get it from GPU enabled priority queue. If none of them return any result, we try to populate the priority queue from the dictionary which gets populated as and when tasks assigned to the workers finish.

Also, please note that the priority queue maintains -(maxAvailableCap) as the queue.PriorityQueue python package only supports min heap.

Started driver locally to confirm everything works fine. README has been updated along with CI as well to run all existing tests in the LoadAware mode as well.

### CODE REVIEW

**Review by aadhithNarayan — DISMISSED** [review]
Can we have a UT just for getWorker flow? It doesn't have to involve gprc calls. 
For eg: Registering a couple of workers(gpu and non gpu) and how getWorker responds with assertions for errors.
It really helps to understand the code better and ensures non breakability.

**Comment by swarnim-vyas** [comment]
> Can we have a UT just for getWorker flow? It doesn't have to involve gprc calls. For eg: Registering a couple of workers(gpu and non gpu) and how getWorker responds with assertions for errors. It really helps to understand the code better and ensures non breakability.

**Comment by swarnim-vyas** [comment]
I will create those UT in a follow up PR

File: Data/WorkerInfo.py [hunk 1]
```diff
@@ -1,10 +1,17 @@
...
         self.portNumber = portNumber
         self.maxThreadCount = maxThreadCount
         self.isGPUEnabled = isGPUEnabled
+        self.hardwareGeneration = hardwareGeneration
 
     def __str__(self):
-        return "[HostName: {} || PortNumber: {} || MaxThreadCount: {} || IsGPUEnabled: {}]".format(self.hostName, self.portNumber, self.maxThreadCount, self.isGPUEnabled)
+        return "[HostName: {} || PortNumber: {} || MaxThreadCount: {} || IsGPUEnabled: {} || Hardware Generation: {}]".format(self.hostName, self.portNumber, self.maxThreadCount, self.isGPUEnabled, self.hardwareGeneration)
+
+    def __lt__(self, other):
```
Comment (krish8484) [inline Data/WorkerInfo.py:13]: what is the purpose of this?

File: scheduler.py [hunk 2]
```diff
@@ -121,4 +150,81 @@ def get_worker(self) -> Union[WorkerInfo, WorkerInfoWrapper]:
             else:
                 logging.info(f"No worker found.")
                 raise NoWorkerAvailableError(self.schedulerMode)
-        
\ No newline at end of file
+        elif self.schedulerMode == constants.SCHEDULINGMODE_LOADAWARE:
+            logging.info(f"Getting worker for {self.schedulerMode}")
+            wrappedWorkerInfo = self.get_worker_info(self.WorkersWithoutGPU, self.WorkersWithoutGPULock)
```
Comment (krish8484) [inline scheduler.py:155]: why do we check WorkersWithoutGPU before WorkersWithGPU? This should depend on the type of the task?

**Review by aadhithNarayan — APPROVED** [review]

File: scheduler.py (hunk 2 above)
Comment (swarnim-vyas) [reply scheduler.py:155]: yeah this depends on type of task - given that isn't available right now, we assume all tasks do not need GPU

File: Data/WorkerInfo.py (hunk 1 above)
Comment (swarnim-vyas) [reply Data/WorkerInfo.py:13]: The lt/gt overwrite is used when in priority queue's first element is same, the internal python implementation uses the second property (in this case workerInfo) to maintain maxHeap/minHeap property

**Review by krish8484 — APPROVED** [review]

**Pull request merged**